W-15501420-Include-ACB-on-Overview-CP (#269)
mulesoft/docs-munit: modules/ROOT/pages/munit-test-concept.adoc

include::_attributes.adoc[]
endif::[]
:page-aliases: munit-suite.adoc, munit-test-reference.adoc
:page-deployment-options: acb-ide, studio-ide

== MUnit Test Suite

> The base file in MUnit is the test suite file, an XML file located in the `src/test/munit` directory of your Mule application project. You can have as many MUnit test suite files as needed. Each MUnit test suite file is a collection of MUnit tests and runs independently from other MUnit test suites.

Reviewer: ```suggestion
The base file in MUnit is the test suite file, an XML file located in the `src/test/munit` directory of your Mule application project. You can use multiple MUnit test suite files. Each MUnit test suite file is a collection of MUnit tests and runs independently from other MUnit test suites.
```

Author: OK! Thanks!

==== Minimum Mule Version

> To specify that a suite must run if the Mule version is equal to or newer than a specific version, use the `minMuleVersion` attribute of the configuration:

Reviewer: ```suggestion
Specify the minimum required Mule version to run a test suite by defining the `minMuleVersion` attribute of the configuration:
```

Author: OK! Thanks!

|Name |Description

|`name`
| Defines the name of the test. The name must be unique in the test suite. +
This attribute is mandatory.

|`description`
|Describes the scenario to test.

|`ignore`
|Defines if the test must be ignored. If not present, the test isn't ignored.

|`tags`
|Comma-separated list of tags to assign to the test.

|`expectedErrorType`
|Defines the error Type to receive after executing the test.

|`expectedException`
|Defines the exception to receive after executing the test.

|`expectedErrorDescription`
|Defines the error description to receive after executing the test.

|`timeOut`
|Defines the maximum time in milliseconds that the test can run before failing.

|===

==== MUnit Test Description Attribute

> In MUnit, you must write a description for your test. The description must be useful and representative of the tested scenario. This description displays in the test console before running the test, and also in the reports. The more representative the description, the easier to read and troubleshoot any failures.

Reviewer: ```suggestion
MUnit tests require a description. The description must be useful and representative of the tested scenario. This description shows in the test console before running the test, and also in the reports. The more representative the description, the easier it is to read and troubleshoot failures.
```

Author: OK! Thanks!

==== Ignore a Test

> You can find scenarios where you must shut down a test. Whether the test is failing or it has inconvenient side effects, you can ignore a specific test so that you don't manually comment out the code.

Reviewer: ```suggestion
When a test fails or has inconvenient side effects, you can shut it down without commenting out the code.
```

Author: OK! Thanks!

> You can ignore any of your tests by adding the `ignore` boolean to the test definition:

Reviewer: ```suggestion
To ignore a test, add the `ignore` boolean to its definition:
```

Author: OK! Thanks!

</munit:test>
----

.In Anypoint Studio, you can define tags in the *Tags* field.
image::munit-test-concept-c2d9f.png["Tags field highlighted in Studio."]

==== MUnit Test Expected Error and Exception Attributes

> If you want to validate that the flow or sub-flow you're testing fails and throws a specific error, you can add the attributes `expectedErrorType`, `expectException`, and `expectedErrorDescription` as follows:

Reviewer: ```suggestion
To validate that a flow or sub-flow you're testing fails and throws a specific error, add the `expectedErrorType`, `expectException`, and `expectedErrorDescription` attributes as follows:
```

Author: OK! Thanks!

----

.A visual representation on Anypoint Studio
image::munit-test-concept-e7280.png["Failure Expectation section highlighted in Studio."]

* The `expectedErrorType` attribute expects an error ID defined in the tested application. +

> With this attribute, you can validate that a defined error type in your application is thrown. If you define an `errorType` that does not exist in your application, the test doesn't run.

Reviewer: ```suggestion
This attribute validates that your application throws a defined error type. If you define an `errorType` that does not exist in your application, the test doesn't run.
```

Author: OK! Thanks!

+
If you define that your test expects an exception and none is thrown, the test fails immediately.

* The attribute `expectedErrorDescription` expects an error description. +

> With this attribute, you can validate that an error is thrown and the error has a description that is equal to or contains the expected error description. If you define that your test expects an error and none is thrown or the error thrown description doesn't match, the test fails immediately.

Reviewer: ```suggestion
This attribute validates that your application throws an error and the error has a description that is equal to or contains the expected error description. If you define that your test expects an error and none is thrown or the error description doesn't match, the test fails immediately.
```

Author: OK! Thanks!
